### PR TITLE
Feature/external api integration

### PR DESCRIPTION
**Pull request by luismiguel0529**
## Summary
- Add folder service
- Add in AppConfig bean RestTemplate
- Add url api in aplication.properties
- Add classOpenLibraryService
- Finished PR structure
## Trello Card
https://trello.com/c/w7IhABYx/22-integraci%C3%B3n-api-externa

### CODE REVIEW

File: src/main/java/wolox/training/config/AppConfig.java [hunk 1]
```diff
@@ -13,4 +16,10 @@ public PasswordEncoder passwordEncoder() {
         return new BCryptPasswordEncoder();
     }
 
+    @Bean
+    public RestTemplate restTemplate(){
+        return new RestTemplate();
+    }
+
+
```
Comment (joredudiaz) [inline src/main/java/wolox/training/config/AppConfig.java]: Can you remove this break line please?

File: src/main/java/wolox/training/models/dto/BookDTO.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    public List<HashMap<String, String>> getAuthors() {
+        return authors;
+    }
+
+    public void setAuthors(List<HashMap<String, String>> authors) {
+        this.authors = authors;
+    }
+
+    public Book setBook() {
+
```
Comment (joredudiaz) [inline src/main/java/wolox/training/models/dto/BookDTO.java]: Can you remove this break line please?

File: src/main/java/wolox/training/service/OpenLibraryService.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        final String isbnQuery = "ISBN:" + isbn;
+        URI uri = UriComponentsBuilder
+                .fromHttpUrl(urlOpenLibrary)
+                .path("books")
+                .queryParam("bibkeys",isbnQuery)
+                .queryParam("format","json")
+                .queryParam("jscmd","data")
+                .build()
+                .toUri();
+
```
Comment (joredudiaz) [inline src/main/java/wolox/training/service/OpenLibraryService.java]: Can you remove this break line please?

File: src/main/java/wolox/training/service/OpenLibraryService.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            bookDTO.setTitle(mapper.convertValue(node.get(isbnQuery).get("title"),String.class));
+            bookDTO.setSubtitle(mapper.convertValue(node.get(isbnQuery).get("subtitle"),String.class));
+            bookDTO.setPublishers(mapper.convertValue(node.get(isbnQuery).get("publishers"),List.class));
+            bookDTO.setPublishDate(mapper.convertValue(node.get(isbnQuery).get("publish_date"), String.class));
+            bookDTO.setNumberOfPages(mapper.convertValue(node.get(isbnQuery).get("number_of_pages"),String.class));
+            bookDTO.setAuthors(mapper.convertValue(node.get(isbnQuery).get("authors"),List.class));
+            return bookDTO;
+        }
+        throw new BookNotFoundException();
+
```
Comment (joredudiaz) [inline src/main/java/wolox/training/service/OpenLibraryService.java:37]: Can you remove this break line please?

File: src/main/java/wolox/training/service/OpenLibraryService.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            BookDTO bookDTO = new BookDTO();
+            bookDTO.setIsbn(isbn);
+            bookDTO.setTitle(mapper.convertValue(node.get(isbnQuery).get("title"),String.class));
+            bookDTO.setSubtitle(mapper.convertValue(node.get(isbnQuery).get("subtitle"),String.class));
+            bookDTO.setPublishers(mapper.convertValue(node.get(isbnQuery).get("publishers"),List.class));
+            bookDTO.setPublishDate(mapper.convertValue(node.get(isbnQuery).get("publish_date"), String.class));
+            bookDTO.setNumberOfPages(mapper.convertValue(node.get(isbnQuery).get("number_of_pages"),String.class));
+            bookDTO.setAuthors(mapper.convertValue(node.get(isbnQuery).get("authors"),List.class));
+            return bookDTO;
+        }
```
Comment (joredudiaz) [inline src/main/java/wolox/training/service/OpenLibraryService.java:36]: Can you add a break line please?

File: src/main/java/wolox/training/service/OpenLibraryService.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+            ObjectMapper mapper = new ObjectMapper();
+            BookDTO bookDTO = new BookDTO();
+            bookDTO.setIsbn(isbn);
+            bookDTO.setTitle(mapper.convertValue(node.get(isbnQuery).get("title"),String.class));
+            bookDTO.setSubtitle(mapper.convertValue(node.get(isbnQuery).get("subtitle"),String.class));
+            bookDTO.setPublishers(mapper.convertValue(node.get(isbnQuery).get("publishers"),List.class));
+            bookDTO.setPublishDate(mapper.convertValue(node.get(isbnQuery).get("publish_date"), String.class));
+            bookDTO.setNumberOfPages(mapper.convertValue(node.get(isbnQuery).get("number_of_pages"),String.class));
+            bookDTO.setAuthors(mapper.convertValue(node.get(isbnQuery).get("authors"),List.class));
```
Comment (joredudiaz) [inline src/main/java/wolox/training/service/OpenLibraryService.java]: You can let this on the DTO

**Review by joredudiaz — COMMENTED** [review]
You didn't test the service and the endpoint that call it. Test it, please 😄

File: src/main/java/wolox/training/service/OpenLibraryService.java (hunk 6 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/service/OpenLibraryService.java]: corrected please validate

File: src/main/java/wolox/training/service/OpenLibraryService.java (hunk 5 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/service/OpenLibraryService.java:36]: corrected please validate

File: src/main/java/wolox/training/service/OpenLibraryService.java (hunk 4 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/service/OpenLibraryService.java:37]: corrected please validate

File: src/main/java/wolox/training/service/OpenLibraryService.java (hunk 3 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/service/OpenLibraryService.java]: corrected please validate

File: src/main/java/wolox/training/models/dto/BookDTO.java (hunk 2 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/models/dto/BookDTO.java]: corrected please validate

File: src/main/java/wolox/training/config/AppConfig.java (hunk 1 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/config/AppConfig.java]: corrected please validate

File: src/main/java/wolox/training/models/dto/BookDTO.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        book.setTitle(this.title);
+        book.setSubtitle(this.subtitle);
+        book.setPublisher(this.publishers.get(0).get("name"));
+        book.setPages(this.numberOfPages);
+        book.setAuthor(this.authors.get(0).get("name"));
+        book.setYear(this.publishDate);
+        return book;
+    }
+
+    public BookDTO setBookDto(ObjectNode node,BookDTO bookDTO,String isbn){
```
Comment (joredudiaz) [inline src/main/java/wolox/training/models/dto/BookDTO.java]: You can do an static method and create a new BookDTO without pass through parameters

**Review by joredudiaz — CHANGES_REQUESTED** [review]
Just that little thing

File: src/main/java/wolox/training/models/dto/BookDTO.java (hunk 7 above)
Comment (luismiguel0529) [reply src/main/java/wolox/training/models/dto/BookDTO.java]: corrected please validate

File: src/main/java/wolox/training/controllers/BookController.java [hunk 8]
```diff
@@ -123,4 +134,16 @@ public void delete(@PathVariable Long id) {
...
+    @GetMapping("/find-by-isbn")
+    public ResponseEntity<Book> findByIsbn(@RequestParam String isbn) {
+        Optional<Book> bookDB = bookRepository.findByIsbn(isbn);
+        if (bookDB.isPresent()) {
+            return new ResponseEntity<>(bookDB.get(), HttpStatus.OK);
+        } else {
+            BookDTO bookDTO = openLibraryService.findInfoBook(isbn);
+            Book book = bookRepository.save(bookDTO.setBook());
+            return new ResponseEntity<>(book, HttpStatus.CREATED);
+        }
```
Comment (facucompe) [inline src/main/java/wolox/training/controllers/BookController.java]:
```
        return bookRepository.findByIsbn(isbn)
                .map(book -> new ResponseEntity<>(bookDB.get(), HttpStatus.OK))
                .orElseGet(() -> {
                    BookDTO bookDTO = openLibraryService.findInfoBook(isbn);
                    Book book = bookRepository.save(bookDTO.setBook());
                    return new ResponseEntity<>(book, HttpStatus.CREATED);
                });
```

Comment (facucompe) [reply src/main/java/wolox/training/controllers/BookController.java]:
```suggestion
        return bookRepository.findByIsbn(isbn)
                .map(book -> new ResponseEntity<>(bookDB.get(), HttpStatus.OK))
                .orElseGet(() -> {
                    BookDTO bookDTO = openLibraryService.findInfoBook(isbn);
                    Book book = bookRepository.save(bookDTO.setBook());
                    return new ResponseEntity<>(book, HttpStatus.CREATED);
                });
```

**Review by facucompe — DISMISSED** [review]

**Review by joredudiaz — DISMISSED** [review]

**Review by joredudiaz — DISMISSED** [review]

**Review by joredudiaz — APPROVED** [review]

**Review by facucompe — APPROVED** [review]

**Pull request merged**